File: clients/kotlin/analytics/src/test/kotlin/com/vapolia/analytics/SpoolTest.kt
package com.vapolia.analytics

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TemporaryFolder
import java.io.File

class SpoolTest {

    @get:Rule
    val folder = TemporaryFolder()

    private fun spool(capacity: Int = 10, file: File = folder.newFile("spool")) = file to Spool(file, capacity)

    private fun pending(name: String, ts: Long = 1_757_500_000_000) = Pending(
        BatchKey(
            "11111111-0000-0000-0000-000011111111",
            Device(platform = "android", country = "FR"),
            """{"plan":"premium"}""",
        ),
        Event(name, ts, mapOf("result" to "win", "moves" to 34.0, "ok" to true)),
    )

    @Test
    fun `round trips events, device context included`() {
        val (_, disk) = spool()
        val items = listOf(pending("app_open"), pending("game_end"))

        disk.save(items)

        assertEquals(items, disk.load())
    }

    @Test
    fun `reading consumes the file`() {
        val (file, disk) = spool()
        disk.save(listOf(pending("app_open")))

        assertEquals(1, disk.load().size)
        assertFalse(file.exists())
        assertTrue(disk.load().isEmpty())
    }

    @Test
    fun `saving nothing clears the file`() {
        val (file, disk) = spool()
        disk.save(listOf(pending("app_open")))

        disk.save(emptyList())

        assertFalse(file.exists())
    }

    @Test
    fun `a truncated file yields nothing and is removed`() {
        val (file, disk) = spool()
        disk.save(listOf(pending("app_open"), pending("game_end")))
        val bytes = file.readBytes()
        file.writeBytes(bytes.copyOf(bytes.size / 2))

        assertTrue(disk.load().isEmpty())
        assertFalse(file.exists())
    }

    @Test
    fun `capacity keeps the oldest events`() {
        val (_, disk) = spool(capacity = 2)

        disk.save(listOf(pending("first"), pending("second"), pending("third")))

        assertEquals(listOf("first", "second"), disk.load().map { it.event.name })
    }
}

File: clients/js/src/__tests__/spool.test.ts
import { describe, expect, it } from 'vitest';

import { Spool } from '../core/spool';
import type { Pending } from '../core/types';
import { MemoryStorage } from './support';

function pending(name: string): Pending {
  return {
    device: { country: 'FR' },
    installId: '11111111-0000-0000-0000-000011111111',
    event: { name, ts: 1_757_500_000_000, props: { result: 'win', moves: 34, ok: true } },
  };
}

describe('Spool', () => {
  it('round trips events, device context included', async () => {
    const spool = new Spool(new MemoryStorage(), 'testsource', 10);
    const items = [pending('app_open'), pending('game_end')];

    await spool.save(items);

    expect(await spool.load()).toEqual(items);
  });

  it('is read once', async () => {
    const storage = new MemoryStorage();
    const spool = new Spool(storage, 'testsource', 10);
    await spool.save([pending('app_open')]);

    expect(await spool.load()).toHaveLength(1);
    expect(storage.items.size).toBe(0);
    expect(await spool.load()).toHaveLength(0);
  });

  it('clears the key when there is nothing to save', async () => {
    const storage = new MemoryStorage();
    const spool = new Spool(storage, 'testsource', 10);
    await spool.save([pending('app_open')]);

    await spool.save([]);

    expect(storage.items.size).toBe(0);
  });

  it('yields nothing for a truncated payload', async () => {
    const storage = new MemoryStorage();
    const spool = new Spool(storage, 'testsource', 10);
    await spool.save([pending('app_open')]);
    const key = [...storage.items.keys()][0]!;
    storage.items.set(key, storage.items.get(key)!.slice(0, 20));

    expect(await spool.load()).toHaveLength(0);
  });

  it('keeps the oldest events when capped', async () => {
    const spool = new Spool(new MemoryStorage(), 'testsource', 2);

    await spool.save([pending('first'), pending('second'), pending('third')]);

    expect((await spool.load()).map((item) => item.event.name)).toEqual(['first', 'second']);
  });
});
